### PR TITLE
(CDAP-16509) Improve StructuredRecord memory footprint

### PR DESCRIPTION
**Pull request by chtyim**
- Added a schema cache for StructuredRecord

### CODE REVIEW

**Review by sagarkapare — APPROVED** [review]

**Comment by chtyim** [comment]
Build https://builds.cask.co/browse/CDAP-DUT7176

**Review by sagarkapare — APPROVED** [review]

File: cdap-api-common/src/main/java/io/cdap/cdap/api/data/format/LRUCache.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+ * This class is thread safe.
+ *
+ * @param <K> type of the key
+ * @param <V> type of the value
+ */
+@ThreadSafe
+final class LRUCache<K, V> {
+
+  private final Map<K, V> entries;
+  private final PriorityQueue<TimedKey<K>> orderedKey;
```
Comment (albertshau) [inline cdap-api-common/src/main/java/io/cdap/cdap/api/data/format/LRUCache.java]: nit: plural (orderedKeys) is easier to read/understand

**Review by sagarkapare — APPROVED** [review]

**Review by albertshau — APPROVED** [review]
Do you know if the extra time spent hashing the schema for each record is noticeable?

If so we could also potentially check if the most recently used schema is actually the same schema instance.

File: cdap-api-common/src/main/java/io/cdap/cdap/api/data/format/LRUCache.java (hunk 1 above)
Comment (chtyim) [reply cdap-api-common/src/main/java/io/cdap/cdap/api/data/format/LRUCache.java]: Fixed

**Comment by chtyim** [comment]
Computing the hash is the same as equals. In the Schema.equals() implementation, it is actually just comparing the schema hash (it's a MD5 hash, so it assumes uniqueness).

**Comment by chtyim** [comment]
And without the schema hash, we can't tell if the new StructuredRecord's schema is the same as the last used one.

**Comment by albertshau** [comment]
by same, I meant actually the same instance, with ==

**Comment by chtyim** [comment]
We could add a special field in the LRUCache to remember the last inserted schema instance. But it seems a little bit hacky to me since it is optimizing for a single schema in the whole JVM use case.

**Pull request merged**